### PR TITLE
ENH: Add parser for Siemens "ASCCONV" text format

### PR DESCRIPTION
**Pull request by moloney**
This is a clean PR after I screwed up the merge in #418 

This includes Matthew Brett's hard work improving the parser (from his 'ascconv-refactor' branch).

### CODE REVIEW

File: nibabel/nicom/ascconv.py [hunk 1]
```diff
@@ -0,0 +1,207 @@
+# emacs: -*- mode: python-mode; py-indent-offset: 4; indent-tabs-mode: nil -*-
+# vi: set ft=python sts=4 ts=4 sw=4 et:
+"""
+Parse the "ASCCONV" meta data format found in a variety of Siemens MR files.
+"""
+import re
+import ast
+from ..externals import OrderedDict
```
Comment (effigies) [inline nibabel/nicom/ascconv.py:8]: I suspect this is included for compatibility, but this has been in the stdlib for all supported versions of Python:

```suggestion
from collections import OrderedDict
```

File: nibabel/nicom/tests/test_ascconv.py [hunk 2]
```diff
@@ -0,0 +1,63 @@
+""" Testing Siemens "ASCCONV" parser
+"""
+
+from os.path import join as pjoin, dirname
+
+import numpy as np
+
+from .. import ascconv
+from ...externals import OrderedDict
```
Comment (effigies) [inline nibabel/nicom/tests/test_ascconv.py:9]:
```suggestion
from collections import OrderedDict
```

File: nibabel/nicom/ascconv.py (hunk 1 above)
Comment (matthew-brett) [reply nibabel/nicom/ascconv.py:8]: Actually - now we require Python 3.6 - aren't all dicts OrderedDicts?

File: nibabel/nicom/ascconv.py [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+            prev_target_type = list
+        else:
+            raise AscconvParseError(
+                'Unexpected LHS element {0}'.format(target))
+    return reversed(atoms)
+
+
+def _create_obj_in(atom, root):
+    """ Create object defined in `atom` in dict-like given by `root`
+
```
Comment (matthew-brett) [inline nibabel/nicom/ascconv.py:104]: May be worth noting (I should have noted) that returns `atom` if there is already something with the same name in `root`.

File: nibabel/nicom/ascconv.py [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+    if obj is not NoValue:
+        return obj
+    obj = atom.obj_type()
+    root[name] = obj
+    return obj
+
+
+def _create_subscript_in(atom, root):
+    """ Create object defined in `atom` at index ``atom.obj_id`` in list `root`
+
```
Comment (matthew-brett) [inline nibabel/nicom/ascconv.py:118]: Returns existing object at index if exists, else entends object with None and appends at index.

File: nibabel/nicom/ascconv.py [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+    root_obj = namespace
+    for el in atoms:
+        prev_root = root_obj
+        if isinstance(el.op, (ast.Attribute, ast.Name)):
+            root_obj = _create_obj_in(el, root_obj)
+        else:
+            root_obj = _create_subscript_in(el, root_obj)
+        if not isinstance(root_obj, el.obj_type):
+            raise AscconvParseError(
+                'Unexpected type for {0} in {1}'.format(el.obj_id, prev_root))
```
Comment (matthew-brett) [inline nibabel/nicom/ascconv.py:157]: Probably worth using 'f' strings at this point.

File: nibabel/nicom/tests/test_ascconv.py (hunk 2 above)
Comment (matthew-brett) [reply nibabel/nicom/tests/test_ascconv.py:9]: Likewise - can we drop OrderedDict?

**Comment by matthew-brett** [comment]
Ok for me to add the fixes from my comments, and merge?

**Comment by effigies** [comment]
Sounds good to me. @moloney, any objections?

File: nibabel/nicom/ascconv.py [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+ASCCONV_RE = re.compile(
+    r'### ASCCONV BEGIN((?:\s*[^=\s]+=[^=\s]+)*) ###\n(.*?)\n### ASCCONV END ###',
+    flags=re.M | re.S)
+
+
+class AscconvParseError(Exception):
+    """ Error parsing ascconv file """
+
+
+class Atom(object):
```
Comment (mgxd) [inline nibabel/nicom/ascconv.py:20]:
```suggestion
class Atom:
```

File: nibabel/nicom/ascconv.py [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+        Key (``obj_type is dict``) or index (``obj_type is list``)
+    """
+
+    def __init__(self, op, obj_type, obj_id):
+        self.op = op
+        self.obj_type = obj_type
+        self.obj_id = obj_id
+
+
+class NoValue(object):
```
Comment (mgxd) [inline nibabel/nicom/ascconv.py:57]:
```suggestion
class NoValue:
```

**Comment by matthew-brett** [comment]
Thanks - nitpicks applied!

File: nibabel/nicom/ascconv.py [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+    atoms = []
+    prev_target_type = default_class  # Placeholder for any scalar value
+    while True:
+        if isinstance(target, ast.Name):
+            atoms.append(Atom(target, prev_target_type, target.id))
+            break
+        if isinstance(target, ast.Attribute):
+            atoms.append(Atom(target, prev_target_type, target.attr))
+            target = target.value
+            prev_target_type = OrderedDict
```
Comment (effigies) [inline nibabel/nicom/ascconv.py:90]:
```suggestion
            prev_target_type = dict
```

File: nibabel/nicom/ascconv.py [hunk 9]
```diff
@@ -0,0 +1,207 @@
...
+    ascconv_str : str
+        The string we are parsing
+    str_delim : str, optional
+        String delimiter.  Typically '"' or '""'
+
+    Returns
+    -------
+    prot_dict : OrderedDict
+        Meta data pulled from the ASCCONV section.
+    attrs : OrderedDict
```
Comment (effigies) [inline nibabel/nicom/ascconv.py:186]:
```suggestion
    prot_dict : dict
        Meta data pulled from the ASCCONV section.
    attrs : dict
```

File: nibabel/nicom/ascconv.py [hunk 10]
```diff
@@ -0,0 +1,207 @@
...
+        Any attributes stored in the 'ASCCONV BEGIN' line
+
+    Raises
+    ------
+    AsconvParseError
+        A line of the ASCCONV section could not be parsed.
+    '''
+    attrs, content = ASCCONV_RE.match(ascconv_str).groups()
+    attrs = OrderedDict((tuple(x.split('=')) for x in attrs.split()))
```
Comment (effigies) [inline nibabel/nicom/ascconv.py:195]:
```suggestion
    attrs = dict(x.split('=') for x in attrs.split())
```

File: nibabel/nicom/ascconv.py [hunk 11]
```diff
@@ -0,0 +1,207 @@
...
+    # Normalize string start / end markers to something Python understands
+    content = content.replace(str_delim, '"""')
+    # Use Python's own parser to parse modified ASCCONV assignments
+    tree = ast.parse(content)
+
+    prot_dict = OrderedDict()
```
Comment (effigies) [inline nibabel/nicom/ascconv.py:201]:
```suggestion
    prot_dict = dict()
```

File: nibabel/nicom/tests/test_ascconv.py (hunk 2 above)
Comment (effigies) [reply nibabel/nicom/tests/test_ascconv.py:9]:
```suggestion
```

File: nibabel/nicom/tests/test_ascconv.py [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+
+DATA_PATH = pjoin(dirname(__file__), 'data')
+ASCCONV_INPUT = pjoin(DATA_PATH, 'ascconv_sample.txt')
+
+
+def test_ascconv_parse():
+    with open(ASCCONV_INPUT, 'rt') as fobj:
+        contents = fobj.read()
+    ascconv_dict, attrs = ascconv.parse_ascconv(contents, str_delim='""')
+    assert attrs == OrderedDict()
```
Comment (effigies) [inline nibabel/nicom/tests/test_ascconv.py:21]:
```suggestion
    assert attrs == {}
```

File: nibabel/nicom/ascconv.py (hunk 1 above)
Comment (effigies) [reply nibabel/nicom/ascconv.py:8]:
```suggestion
```

**Comment by effigies** [comment]
> Ok for me to add the fixes from my comments, and merge?

@matthew-brett Just saw this (again, probably...). Feel free to patch if you have time. Otherwise I can.

**Comment by matthew-brett** [comment]
I can do the changes.  I think we'll need OrderedDict unless we are dropping Python 3.6.   Is that imminent?

**Comment by effigies** [comment]
No, we only just dropped 3.5.

**Comment by matthew-brett** [comment]
On [further investigation](https://stackoverflow.com/questions/39980323/are-dictionaries-ordered-in-python-3-6), dicts are insertion-ordered in CPython as of CPython 3.6, and guaranteed to be so ordered from Python 3.7.   PyPy actually [had this before CPython](https://morepypy.blogspot.com/2015/01/faster-more-memory-efficient-and-more.html), so I believe Python dicts are insertion ordered in PyPy 3.6. 

So our only worry for 3.6 would be some implementation other than CPython or PyPy, that has chosen some other algorithm for dicts, or kept the old CPython one for some reason.   That seems very unlikely.  So I think we can reasonably drop the requirement for OrderedDict soon.

But - maybe not with this PR.  Maybe that should be a separate change.

What do y'all think?

**Comment by matthew-brett** [comment]
I put some more edits from the comments here into https://github.com/nipy/nibabel/pull/935

Sorry - I should have applied the new commits here, but I forgot how to do that.

**Comment by effigies** [comment]
Hi Matthew. #896 looks identical to this. Did you push all of your commits?

**Comment by matthew-brett** [comment]
Oops - done.

**Pull request closed without merging**